Merge 21.0 (21.0.0) code freeze into `trunk`

mokagio opened the pull request:
- [x] Update internal Pods to stable version
- [x] New version header in `RELEASE-NOTE.txt`
- [x] `Localizable.strings` updated
- [x] `release_notes.txt` updated with notes from `RELEASE-NOTE.txt` for current version
- [x] Version update in `.xcconfig`

mokagio committed:
Bump version number

mokagio committed:
Release Notes: add new section for next version (21.1)

mokagio committed:
Update draft release notes for 21.0.

mokagio committed:
Update draft release notes for 21.0.

mokagio committed:
Remove Jetpack-only entries from WordPress 21.0 raw release notes

mokagio committed:
Remove "Jetpack-only" prefix from Jetpack 21.0 release notes

mokagio committed:
Update internal pods to latest stable versions

Notice the Sentry update, which happened because it is a Tracks
dependency.

For the first time ever, running `bundle exec pod update ...` printed
this kind of warning:

```
[!] `<PBXResourcesBuildPhase UUID=`FABB1FAA2602FC2C00C8785C`>` attempted to initialize an object with an unknown UUID. `FEC3B81A26C2915A00A395C7` for attribute: `files`. This can be the result of a merge and the unknown UUID is being discarded.
```

mokagio committed:
Refresh project file after CocoaPods warning

Running `bundle exec pod update ...` in the previous commit,
b2b297c979465c14814ddaed5c5b636e7f08160d, printed a warning:

```
[!] `<PBXResourcesBuildPhase UUID=`FABB1FAA2602FC2C00C8785C`>` attempted to initialize an object with an unknown UUID. `FEC3B81A26C2915A00A395C7` for attribute: `files`. This can be the result of a merge and the unknown UUID is being discarded.
```

CocoaPods suggests there might have been a merge conflict not resolved
properly in the project file, which is something bound to happen from
time to time in a project with as many contributors as ours—and because
of the inconvenient way Xcode generates and manages the file.

Whenever I run into that kind of issue on my end, I add a new file then
remove it. That procedure is enough to make Xcode "refresh" the project
file, which usually gets rid of any dead reference, like the one
CocoaPods highlighted. And that's exactly what this commit tracks.

I also verified this changed by running the tests.

mokagio committed:
Ensure `differenceLabel` is a localized strings literals

`bundle exec fastlane complete_code_freeze` failed with the following
error, thrown as part of the `generate_strings_file_for_glotpress` lane:

```
[13:20:21]: genstrings: error: bad entry in file WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift (line = 58): Argument is not a literal string.
[13:20:21]: fastlane finished with errors

[!] genstrings: error: bad entry in file WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift (line = 58): Argument is not a literal string.
```

mokagio committed:
Update strings for localization

mokagio commented on WordPress/WordPress.xcodeproj/project.pbxproj:
```diff
@@ -18418,7 +18416,6 @@
 				17039226282E6D2F00F602E9 /* ViewsVisitorsLineChartCell.xib in Resources */,
 				FABB28472603067C00C8785C /* Launch Screen.storyboard in Resources */,
 				F465978F28E65F8A00D5F49A /* celadon-on-white-icon-app-60@2x.png in Resources */,
-				FEC3B81A26C2915A00A395C7 /* SingleButtonTableViewCell.xib in Resources */,
```
The changes in this file clean up what was most likely an incorrect merge on the project file.

See 3fc0ac0a67158c180ce263208c81af8242430640.

mokagio commented on WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift:
```diff
@@ -54,14 +54,34 @@ struct StatsSegmentedControlData {
-            difference.abbreviatedString(),
-            differencePercent.abbreviatedString()
-        )
+        // We want to show something like "+1.2K (5%)" if we have a percentage difference and "1.2K" if we don't.
+        // Because localized strings need to be strings literal, we cannot embed any conditional logic in the `localizedString...` call.
+        // We therefore need to generate different string literals base on the state.
+        let differenceSign = difference < 0 ? "" : "+"
+        if differencePercent != 0 {
+            let stringFormat = NSLocalizedString(
+                "insights.visitorsLineChartCell.differenceLabelWithPercentage",
+                value: "%@%@ (%@%%)",
+                comment: "Difference label for Insights Overview stat, indicating change from previous period, including percentage value. Example: +99.9K (5%)"
+            )
+            return String.localizedStringWithFormat(
+                stringFormat,
+                differenceSign,
+                difference.abbreviatedString(),
+                differencePercent.abbreviatedString()
+            )
+        } else {
+            let stringFormat = NSLocalizedString(
+                "insights.visitorsLineChartCell.differenceLabelWithoutPercentage",
+                value: "%@%@",
+                comment: "Difference label for Insights Overview stat, indicating change from previous period. Example: +99.9K"
+            )
+            return String.localizedStringWithFormat(
+                stringFormat,
+                differenceSign,
+                difference.abbreviatedString()
+            )
+        }
```
While I was at it, I also added reverse-DNS keys to that we can avoid ambiguous translations. See #19028.

FYI @staskus, as per [this commit](https://github.com/wordpress-mobile/WordPress-iOS/pull/19400/commits/8545f1cce030165e5eb88635fcbb063a40bc9d05).

Also, I noticed there is no unit test for this logic. It would be great to add some, even though they wouldn't have caught this issue in particular.

staskus replied:
@mokagio 

Thank you for noticing the issue and making the fixes! 🤝

Should I add additional unit tests and additional cases that @AliSoftware was mentioning as a part of this PR?

AliSoftware commented on Podfile:
```diff
@@ -47,7 +47,7 @@ def wordpress_ui
 
 def wordpress_kit
-  pod 'WordPressKit', '~> 4.58.1-beta.1'
+  pod 'WordPressKit', '~> 4.58', '>= 4.58.1'
```
Any reason for that change?

Why not keep the `~>` notation, to at least avoid allowing an hypothetical `5.x.x` version to break things?

Moreover, why not just keep `~> 4.58` then use `bundle update WordPressKit` to update the used version to `4.58.1` in the lockfile while still respect and allow semantic versioning? `>= 4.58.1` feels too loose and open to breaking changes imho

mokagio replied:
> Why not keep the ~> notation

@AliSoftware from the way your question is worded, it seems like my diff removed the `~> 4.58` requirement. However, it did not. It adds `>= 4.58.1` to it. 

My rationale for adding `>= 4.58.1` is to ensure that we always get _at least_ `4.58.1`. I did that because I think using `~> 4.58` could potentially result in a resolution of `4.58.0` is another dependency has a hard requirement on that version.

(We need `4.58.1` because of a bugfix).

`'~> 4.58', '>= 4.58.1'` will avoid allowing an hypothetical `5.x.y` version to break things. I checked this by using `'~> 1.5', '>= 1.5.1'`, which resolved to `1.8.0` in https://github.com/wordpress-mobile/WordPress-iOS/commits/5db3cf6fe3be33d7346ed6a21e03c63e57a3a189 .

Does this address your concern? 

---

Assuming I got the configuration right, what I like about it is the granularity and how it specifies to a reader that there's something "special" about `4.58.1` in relationship to the code.

At the same time,  I can see this being confusing for people who don't spend much time tinkering with dependency managers. 🤔 A single `~> MAJOR.MINOR` would definitely be simpler. And, at the end of the day, we don't have many internal libraries and transitive dependencies at this point, and we might be well served by a simpler setup.

AliSoftware replied:
> from the way your question is worded, it seems like my diff removed the ~> 4.58 requirement. However, it did not. It adds >= 4.58.1 to it.

Oh, good point, I totally misread that diff! ☕  Thanks for the precision 🙂 

In that case it all makes sense 👍

mokagio replied:
Oh, and the reason I didn't do `~> 4.58.1` is because that would resolve only in versions compatible with `4.58.1`. That is, `4.58.2`, `4.58.3`, `4.58.99999`, but not `4.59.0`.

AliSoftware replied:
Yep, totally 🙂  Good thinking 👍 

(For the record, I wasn't suggesting to do `~> 4.58.1`, but just to keep `~> 4.58` and rely on `pod update` to pick `4.58.1`… but now that I see that your `>= 4.58.1` is an addition not a replacement, this all fits and covers all the cases we want indeed 👍

AliSoftware commented on WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift:
```diff
@@ -54,14 +54,34 @@ struct StatsSegmentedControlData {
-        )
+        // We want to show something like "+1.2K (5%)" if we have a percentage difference and "1.2K" if we don't.
+        // Because localized strings need to be strings literal, we cannot embed any conditional logic in the `localizedString...` call.
+        // We therefore need to generate different string literals base on the state.
```
```suggestion
        // We therefore need to generate different string literals based on the state.
```

AliSoftware commented on WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift:
```diff
@@ -54,14 +54,34 @@ struct StatsSegmentedControlData {
-            differencePercent.abbreviatedString()
-        )
+        // We want to show something like "+1.2K (5%)" if we have a percentage difference and "1.2K" if we don't.
+        // Because localized strings need to be strings literal, we cannot embed any conditional logic in the `localizedString...` call.
```
👍  Nice catch!

Not sure how that was missed in the original PR  😓, did `gentrings` warned you about it during code freeze, telling you that the parameters have to be static values? If so, maybe we should run `genstrings` on CI on each PR (even if we don't commit the generated `Localizable.strings` on those PR CI runs, but just to gather potential warnings generated by it)? 🤔

mokagio replied:
> did `gentrings` warned you about it during code freeze, telling you that the parameters have to be static values?

Yes, _phew_. The error [was](https://github.com/wordpress-mobile/WordPress-iOS/pull/19472/commits/65bcdd4c02a5e3a3b17efc6405f00c29b377de72):

```
[!] genstrings: error: bad entry in file WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift (line = 58): Argument is not a literal string.
```

> If so, maybe we should run `genstrings` on CI on each PR (even if we don't commit the generated `Localizable.strings` on those PR CI runs, but just to gather potential warnings generated by it)? 🤔

Yes! I thought about that too but didn't flesh out how it could work. I'd like a dedicated step, so the failure message is clear, but I worry if it's a waste to spin up a macOS box etc. "just" for that.

AliSoftware replied:
> but I worry if it's a waste to spin up a macOS box etc. "just" for that.

Fair point. Maybe we could integrate that in an existing step (e.g. the one already doing `swiftlint`)?

AliSoftware commented on WordPress/Classes/ViewRelated/Stats/Insights/ViewsVisitors/ViewsVisitorsLineChartCell.swift:
```diff
@@ -54,14 +54,34 @@ struct StatsSegmentedControlData {
+        // We want to show something like "+1.2K (5%)" if we have a percentage difference and "1.2K" if we don't.
+        // Because localized strings need to be strings literal, we cannot embed any conditional logic in the `localizedString...` call.
+        // We therefore need to generate different string literals base on the state.
+        let differenceSign = difference < 0 ? "" : "+"
```
Even that conditional about `differenceSign` is making assumptions about the fact that those two variants would always be translated the same with just the added `+` in all languages — while depending on how different locales decide to display those values according to locale rules and practices, that might not be true (e.g. either they might put the `+` sign elsewhere like after instead of before in some locales, or they might have some other character for those cases, just like Arabic has a different Unicode character for the % sign, aka U+066A, or something else…)

So personally I'd even suggest to provide 4 different strings and not just 2, to cover all cases while allowing polyglots to still provide all variants matching their locale rules.

(See also: pbMoDN-3tH-p2#the-tips)

staskus replied:
Good point, we shouldn't be assuming particular symbols or their order in different languages.

mokagio replied:
Good point. On it...

AliSoftware reviewed (changes requested):
- [x] Pods updated to stable version
   - ⚠️  See inline comment about `>=`
- [x] New header in `RELEASE-NOTES.txt`
- [x] JP and WP `release_notes.txt` updated with draft notes
- [x] JP release notes audited for differences between WP and JP
    - [x] Changes from 19383, 19378 and 19414 only kept in JP release notes and removed from WP 👍 
- [x] `WordPress/Resources/en.lproj/Localizable.strings` updated with latest strings
- [x] Version bump in `.xcconfig` files

---

Still requested changes because I think we shouldn't use `>=` in the `Podfile` and keep `~>` instead, but also because I think the strings for the Insights needs to be split even further, in 4 strings instead of 2. And if we do so, better do it before sending those strings to polyglots for translation.

mokagio committed:
Fix a typo in a comment

Co-authored-by: Olivier Halligon <olivier.halligon@automattic.com>

mokagio committed:
Tweak a localization to remove any assumption on "+" localization

See discussion at
https://github.com/wordpress-mobile/WordPress-iOS/pull/19472#discussion_r996700821

mokagio committed:
Update strings for localization

AliSoftware commented:
> mokagio requested a review from AliSoftware 10 minutes ago

I think we need to wait for https://github.com/wordpress-mobile/WordPress-iOS/pull/19473 to land first before approving and (auto-)merging this PR to `trunk`, as we don't want the current keys to be sent to GlotPress and to start being localized by poyglots… if we're going to change them via #19473 right after (which will require us to land it to `trunk` again anyway).

So I'll wait for that other PR to land before updating my review of this one; but for the record, apart from that pending fix, the rest LGTM 👍

mokagio committed:
Refine `differenceLabel` implementation and add tests

mokagio committed:
Update strings for localization

mokagio committed:
Improve Insight `differenceLabel` localization logic (#19473)

AliSoftware reviewed (approved):
Approving now that the tweak to fix L10n issue has been merged.

mokagio commented:
Thank you for following up @AliSoftware
